subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileCollectionVisitingSnapshotBuilder.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import java.util.List;
+
+/**
+ * Used to build a {@link FileCollectionSnapshot} by collecting normalized file snapshots.
+ */
+public class FileCollectionVisitingSnapshotBuilder implements FileVisitingSnapshotBuilder {
```

Comment:
I don't like the name very much. This is still a `FileSnapshotVisitor`. And it is also a `FileCollectionSnapshotBuilder`. The type for this is currently called `FileVisitingSnapshotBuilder`. Moreover, `FileCollectionSnapshotBuilder` seems to be better suited for an interface with the `build` method.
How about we do the following renames:
- `FileVisitingSnapshotBuilder` -> `VisitingFileCollectionSnapshotBuilder`
- `FileCollectionSnapshotBuilder` -> `CollectingFileCollectionSnapshotBuilder`
- We create an interface `FileCollectionSnapshotBuilder` with only the method `FileCollectionSnapshot build()`
- `FileCollectionVisitingSnapshotBuilder` -> `GenericFileCollectionSnapshotBuilder`

WDYT?